lib/spack/spack/test/concretize.py

```diff
@@ -103,16 +102,17 @@ def current_host(request, monkeypatch):
         monkeypatch.setattr(spack.platforms.test.Test, 'default', cpu)
         yield target
     else:
-        # There's a cache that needs to be cleared for unit tests
-        PackagePrefs._packages_config_cache = None
         with spack.config.override('packages:all', {'target': [cpu]}):
             yield target
 
     # clear any test values fetched
     spack.architecture.get_platform.cache.clear()
 
 
-@pytest.mark.usefixtures('config', 'mock_packages')
+# This must use the mutable_config fixture because the test
+# adjusting_default_target_based_on_compiler uses the current_host fixture,
+# which changes the config.
+@pytest.mark.usefixtures('mutable_config', 'mock_packages')
```

Comment:
Can we keep `config` here and use `mutable_config`  only in `adjusting_default_target_based_on_compiler`?

Reply:
> Any test using config.override should use mutable_config (since this can change the contents of the in-memory config cache)

As a side comment, in `mock_store` we change permissions of all the relevant files to be read-only so that a test that needs a `mutable_database` but uses instead a `database` would fail. I was wondering if the same can be done for config files.

Reply:
That's possible although there are cases where config could end up writing new files/scopes, which would result in the override without violating file permissions. I should think on it.

Reply:
Regarding https://github.com/spack/spack/pull/16003/files#r412901893

I wasn't aware of how to "override" a fixture for a single test (i.e. use `mutable_config` instead of config). Or are you suggesting I could extract it into a different testing class?

Reply:
I was thinking of doing:
```python
@pytest.mark.usefixtures('config', 'mock_packages')
class TestConcretize(object):
    ...
    @pytest.martk.usefixtures('mutable_config')
    def test_adjusting_default_target_based_on_compiler
```
This means that every test in the class uses the `config` fixture, while `test_adjust_*` uses `mutable_config`.

Reply:
I tried that and it didn't work. I'm not sure if that can be done from reading through https://docs.pytest.org/en/latest/fixture.html, even if it could be done, I'm not sure if it could be done in this way.

For example I think what this means is to use the `config` *and* `mutable_config` fixtures. Depending on the execution order this could have strange results.

Reply:
I just double-checked and you're correct. I was erroneously convinced that `mutable_config` was building on `config` as `mutable_database` is building on `database`, but that's not the case. 

If `mutable_config` dependend on `config` what I suggested above should have worked. Here I agree that it's at confusing if not plain wrong since there's no clear ordering for the fixture setup. 

I'm fine with the current state of the code. Another option if we want to be more specific is to remove `mutable_config` from the class level and replicate either `config` or `mutable_config` for each method (but that might be an overkill).